### PR TITLE
Move `ATen/core/DimVector.h` to `c10/util/DimVector.h`.

### PR DESCRIPTION
**Pull request by ysiraichi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #76098
* #76097
* #76096
* __->__ #77045
* #76812

This PR moves `ATen/core/DimVector.h`, as suggested in:
https://github.com/pytorch/pytorch/pull/76812#discussion_r866875924

The changes can be summarized as:

- Changing includes from `ATen/core/DimVector.h` to `c10/util/DimVector.h`
- Re-declaring both the type and constant size in `at` namespace
- Making `c10::contiguous_strides` return a `DimVector`

### CODE REVIEW

**Comment by ezyang** [comment]
lint

**Review by peterbell10 — CHANGES_REQUESTED** [review]

File: aten/src/ATen/ExpandUtils.h [hunk 1]
```diff
@@ -6,7 +6,7 @@
 #include <ATen/ops/view_copy.h>
 #endif
 
-#include <ATen/core/DimVector.h>
+#include <ATen/DimVector.h>
```
[the remark this answers is not in the document]
Comment (ysiraichi) [reply aten/src/ATen/ExpandUtils.h:9]: Agreed. Also, that would reduce the number of modified files.

**Review by peterbell10 — APPROVED** [review]

File: c10/util/strides.h [hunk 2]
```diff
@@ -1,15 +1,15 @@
...
-    const IntArrayRef sizes) {
+static inline DimVector contiguous_strides(const IntArrayRef sizes) {
   using Int = IntArrayRef::value_type;
   const Int dims = static_cast<Int>(sizes.size());
 
-  std::vector<Int> strides;
+  DimVector strides;
 
   if (dims > 0) {
     strides.assign(dims, 0);
```
Comment (Lezcano) [inline c10/util/strides.h:15]: I'd like to take this failed landing to small improvement to this code:
```suggestion
  DimVector strides(dims);

  if (dims > 0) {
```

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**